# 1 Data wrangling<a id='1_Data_wrangling'></a>

## 1.1 Contents<a id='1.1_Contents'></a>
* [1 Data wrangling](#1_Data_wrangling)
  * [1.1 Contents](#1.1_Contents)
  * [1.2 Introduction](#1.2_Introduction)
    * [1.2.1 Recap Of Data Science Problem](#1.2.1_Recap_Of_Data_Science_Problem)
    * [1.2.2 Introduction To Notebook](#1.2.2_Introduction_To_Notebook)
  * [1.3 Imports](#1.3_Imports)
  * [1.4 Objectives](#1.4_Objectives)
  * [1.5 Load Data](#1.5_Load_Data)
  * [1.6 Explore The Data](#1.6_Explore_The_Data)
  * [1.7 Summary](#1.7_Summary)

## 1.2 Introduction<a id='1.2_Introduction'></a>

This step focuses on collecting your data, organizing it, and making sure it's well defined. Some data cleaning will be done at this stage, but mostly focus is on exploring the data to better understand it.

### 1.2.1 Recap Of Data Science Problem<a id='1.2.1_Recap_Of_Data_Science_Problem'></a>

The purpose of this data science project is to come up with a model for energy production and consumption prediction based on data collected from 1980 to 2021. Energy prediction is a data driven approach which can help understanding how much production and consumption increasing or decreasing over the time. It involves gathering and analyzing data related to energy resources to derive insights and make informed decisions. Main objective of this analysis to make of prediction about rate of energy production and consumption. This model will be used to provide guidance for enery companird about production and demand.

### 1.2.2 Introduction To Notebook<a id='1.2.2_Introduction_To_Notebook'></a>

In this notebook, I will use well structured, helpful headings that frequently are self-explanatory, and make a brief note after any results to highlight key takeaways. This is an immense help to anyone reading your notebook and it will greatly help you when I come to summarise your findings. Note down key findings in a final summary at the end of the notebook. This is a great way to ensure important results don't get lost in the middle of  notebooks.

## 1.3 Imports<a id='2.3_Imports'></a>

In [39]:
#Import pandas, matplotlib.pyplot, and seaborn
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os
import glob


## 1.4 Objectives<a id='1.4_Objectives'></a>

There are some fundamental questions to resolve in this notebook.

* Do you think you may have the data you need to tackle the desired question?
    * Do you have potentially useful features?
* Do you have any fundamental issues with the data?

## 1.5 Load Data<a id='1.5_Load_Data'></a>

First steps in auditing the data are the info method and displaying the first few records with head.

In [40]:
# this is how to get the list of file names according to wild-card expression below
list_of_files = glob.glob("../Data/" + "/*.csv")
list_of_files

['../Data/Production_Nuclear+renewables.csv',
 '../Data/Production_Coal.csv',
 '../Data/Consumption_Neuclear+renewables.csv',
 '../Data/Consumption_Petroleum.csv',
 '../Data/Consumption_Coal.csv',
 '../Data/Production_Pertroleum.csv',
 '../Data/Consumption_NaturalGas.csv',
 '../Data/Production_NaturalGas.csv']

There are 8 csv files for production and consumption of different energy sources: Coal, Neualear, Pertroleum, NaturalGas. First I will take intial look at data to understand how it is structured.

In [41]:
for the_file in list_of_files:
    print(the_file)
    df = pd.read_csv(the_file)
    print(df.head())
    print(df.columns)
    

../Data/Production_Nuclear+renewables.csv
  Continent               Country         1980         1981         1982  \
0    Africa               Algeria  0.002576225  0.003783987  0.004955198   
1    Africa                Angola  0.005505642  0.005540091  0.006272403   
2    Africa                 Benin            0            0            0   
3    Africa              Botswana            0            0            0   
4    Africa          Burkina Faso            0            0            0   

          1983        1984         1985         1986         1987  ...  \
0  0.002451164  0.00538704  0.006675633  0.002580162  0.005115729  ...   
1  0.006312011    0.006264    0.0062682   0.00689436   0.00687654  ...   
2            0           0            0            0            0  ...   
3            0           0            0            0            0  ...   
4            0           0            0            0            0  ...   

          2013        2014         2015         2016    

All the files are in same format. Have columns:['Continent', 'Country', '1980', '1981', '1982', '1983', '1984', '1985',
       '1986', '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994',
       '1995', '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003',
       '2004', '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012',
       '2013', '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021',
       'Total Energy']
       
Last column is sum of an energy type for all the years for a paritucular country in an continent. I will verify that last column is total energy for one file because all the files have same format.

       

In [ ]:
for the_file in list_of_files:
    print(the_file)
    
    coal_df['dist'] = ""
# calculating difference
coal_df['distance'] = abs(coal_df['Consumption_Coal']-coal_df['sum_energy'])
# defining mismatch tolerance
tolerance = 0.00001
#checking match by looping through all the values in dataframe
for i in coal_df['distance']:
    if i <= tolerance:
        coal_df['dist'] = "Match"
    else:
        coal_df['dist'] = "Not match"

In [4]:
# loading coal conusmption file
coal_df = pd.read_csv('../Data/Consumption_Coal.csv')

In [5]:
# print head of coal consumption dataset
coal_df.head()

Continent               Country         1980         1981         1982  \
0    Africa               Algeria  0.002547398  0.007655064  0.020475801   
1    Africa                Angola            0            0            0   
2    Africa                 Benin            0            0            0   
3    Africa              Botswana  0.008660526  0.009356617  0.010099701   
4    Africa          Burkina Faso            0            0            0   

          1983         1984         1985         1986         1987  ...  \
0  0.033213824  0.034235723  0.034242132  0.034610211  0.035079318  ...   
1            0            0            0            0            0  ...   
2            0            0            0            0            0  ...   
3   0.00986717  0.009751175  0.010819193   0.01211937  0.012351123  ...   
4            0            0            0            0            0  ...   

          2013         2014         2015         2016         2017  \
0  0.001259621  0.000835328  0.000576922  0.000575219  0.000924225   
1            0            0            0            0            0   
2     3.17E-05  0.000961001  0.001037911  0.001985751  0.002992505   
3  0.014232074  0.016837001  0.027546501  0.038035577  0.035819574   
4            0            0            0            0            0   

          2018         2019         2020         2021 Consumption_Coal  
0  0.000410418  0.000405436  0.000658437   0.00094109         0.594818  
1            0            0            0            0         0.000000  
2  0.002987991  0.002938642  0.001951128  0.003932631         0.019046  
3  0.033554487  0.031321952  0.029050749  0.032820821         0.808603  
4            0            0            0            0         0.000000  

[5 rows x 45 columns]

In [6]:
# printing data type of all the columns in coal consumption dataset
coal_df.dtypes

Continent            object
Country              object
1980                 object
1981                 object
1982                 object
1983                 object
1984                 object
1985                 object
1986                 object
1987                 object
1988                 object
1989                 object
1990                 object
1991                 object
1992                 object
1993                 object
1994                 object
1995                 object
1996                 object
1997                 object
1998                 object
1999                 object
2000                 object
2001                 object
2002                 object
2003                 object
2004                 object
2005                 object
2006                 object
2007                 object
2008                 object
2009                 object
2010                 object
2011                 object
2012                 object
2013                

Columns for all the years are object datatype. To calculate its sum, first I will convert years columns to numeric.

In [24]:
# creating a list of year columns and converting it to numeric
cols = ['1980', '1981', '1982', '1983', '1984', '1985',
       '1986', '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994',
       '1995', '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003',
       '2004', '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012',
       '2013', '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021']
#coal_df[cols] = coal_df[cols].apply(pd.to_numeric, errors='coerce')

In [78]:
# check data type of coal consumption dataset to verify year columns converted to numeric
coal_df.dtypes

Continent            object
Country              object
1980                float64
1981                float64
1982                float64
1983                float64
1984                float64
1985                float64
1986                float64
1987                float64
1988                float64
1989                float64
1990                float64
1991                float64
1992                float64
1993                float64
1994                float64
1995                float64
1996                float64
1997                float64
1998                float64
1999                float64
2000                float64
2001                float64
2002                float64
2003                float64
2004                float64
2005                float64
2006                float64
2007                float64
2008                float64
2009                float64
2010                float64
2011                float64
2012                float64
2013                

In [79]:
# calulating sum of energy in the years
coal_df['sum_energy'] = coal_df[cols].sum(axis=1)

In [80]:
# checking sum of energy column added to coal consumption dataset
coal_df.head()

Continent               Country      1980      1981      1982      1983  \
0    Africa               Algeria  0.002547  0.007655  0.020476  0.033214   
1    Africa                Angola  0.000000  0.000000  0.000000  0.000000   
2    Africa                 Benin  0.000000  0.000000  0.000000  0.000000   
3    Africa              Botswana  0.008661  0.009357  0.010100  0.009867   
4    Africa          Burkina Faso  0.000000  0.000000  0.000000  0.000000   

       1984      1985      1986      1987  ...      2014      2015      2016  \
0  0.034236  0.034242  0.034610  0.035079  ...  0.000835  0.000577  0.000575   
1  0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   
2  0.000000  0.000000  0.000000  0.000000  ...  0.000961  0.001038  0.001986   
3  0.009751  0.010819  0.012119  0.012351  ...  0.016837  0.027547  0.038036   
4  0.000000  0.000000  0.000000  0.000000  ...  0.000000  0.000000  0.000000   

       2017      2018      2019      2020      2021  Consumption_Coal  \
0  0.000924  0.000410  0.000405  0.000658  0.000941          0.594818   
1  0.000000  0.000000  0.000000  0.000000  0.000000          0.000000   
2  0.002993  0.002988  0.002939  0.001951  0.003933          0.019046   
3  0.035820  0.033554  0.031322  0.029051  0.032821          0.808603   
4  0.000000  0.000000  0.000000  0.000000  0.000000          0.000000   

   sum_energy  
0    0.594819  
1    0.000000  
2    0.019046  
3    0.808603  
4    0.000000  

[5 rows x 46 columns]

In [81]:
# Compare calculated sum of energy to total energy column in dataframe

# creating new column with empty value to save value of matching calculated sum of energy to total energy column
coal_df['dist'] = ""
# calculating difference
coal_df['distance'] = abs(coal_df['Consumption_Coal']-coal_df['sum_energy'])
# defining mismatch tolerance
tolerance = 0.00001
#checking match by looping through all the values in dataframe
for i in coal_df['distance']:
    if i <= tolerance:
        coal_df['dist'] = "Match"
    else:
        coal_df['dist'] = "Not match"


In [82]:
coal_df['dist'].value_counts()

Match    230
Name: dist, dtype: int64

In [6]:
# creating a list of year columns and converting it to numeric
cols = ['1980', '1981', '1982', '1983', '1984', '1985',
       '1986', '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994',
       '1995', '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003',
       '2004', '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012',
       '2013', '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021']


In [7]:
for the_file in list_of_files:
    df = pd.read_csv(the_file)
    df[cols] = df[cols].apply(pd.to_numeric, errors='coerce')
    df['sum_energy'] = df[cols].sum(axis=1)
    print(df.columns)
    
    df['dist'] = ""
    # calculating difference
    total = df.iloc[:,-2:]
   # print(last_col)
    tolerance = 0.00001
    #if abs(df.iloc[:,-2:]-df['sum_energy']) <= tolerance:
     #   continue
    df['distance'] = df[abs(df.iloc[:,-2:]-df['sum_energy'])]
    # defining mismatch tolerance
 
    #checking match by looping through all the values in dataframe
    for i in df['distance']:
        if i <= tolerance:
            df['dist'] = "Match"
        else:
            df['dist'] = "Not match"


Index(['Continent', 'Country', '1980', '1981', '1982', '1983', '1984', '1985',
       '1986', '1987', '1988', '1989', '1990', '1991', '1992', '1993', '1994',
       '1995', '1996', '1997', '1998', '1999', '2000', '2001', '2002', '2003',
       '2004', '2005', '2006', '2007', '2008', '2009', '2010', '2011', '2012',
       '2013', '2014', '2015', '2016', '2017', '2018', '2019', '2020', '2021',
       'Production_Neuclear', 'sum_energy'],
      dtype='object')


ValueError: Cannot set a DataFrame with multiple columns to the single column distance

As shown above sum of evergy in all the years is equal to total energy, I can exclude this column while creating final dataframe because I will be analyzing how a particular energy production or consumption changes over the years. To do this, I need all the years in one column and its energy values in the another column. I will be applying melt funtion and keeping total energy column will complicate data.

### Combine multiple files

In [42]:
# Defining dataframe that I want afting combining all the files
final_df = pd.DataFrame(columns=["Continent", "Country", "Year", "Energy", "Usage", "Source"])

#loop through all the files
for the_file in list_of_files:

    df = pd.read_csv(the_file)
    # add all the column names in the list
    df_col_list = df.columns.tolist()
    
    #checking number of rows in dataframe
    (n_rows, n_cols) = df.shape
    print(n_rows)
    # to get usage of energy (production or consumption) or type of energy, take value from last column name
    Usage_type = df_col_list[-1].split("_")[0]
    Energy_type = df_col_list[-1].split("_")[1]
    
    # Remove total energy column
    df = df.loc[:, df.columns!=df_col_list[-1]]
    # melting data to get all the years in onve column
    df_melt = df.melt(id_vars=['Continent', 'Country'],
                   var_name='Year',
                   value_name='Energy')
    
    # Notice that for each row of `df`, we would get as many rows of the melted DF as there are years in `df`; 
    #and this number of years is `len(the_cols)` minus 3, which is the number of columns in `df` that are not years.
    #In conclusion, the number of rows of the melted DF should be the number of rows of `df` times the number of years, 
    #which is `len(the_cols)` minus 3
    print(df_melt.shape[0] == n_rows * (len(df_col_list) - 3))
        
    
    # defining columns for usage (production or consumption) and type of energy
    df_melt['Usage'] = Usage_type
    df_melt['Source'] = Energy_type
    #concatinating dataframe to get one dataframe
    final_df = pd.concat([final_df, df_melt], ignore_index=True, axis=0)
    

230
True
230
True
230
True
230
True
230
True
230
True
230
True
230
True


## 1.6 Explore The Data<a id='1.6_Explore_The_Data'></a>


In [43]:
# checking shape of final dataframe
final_df.shape

(77280, 6)

In [44]:
# printing head
final_df.head()

Continent               Country  Year       Energy       Usage    Source
0    Africa               Algeria  1980  0.002576225  Production  Neuclear
1    Africa                Angola  1980  0.005505642  Production  Neuclear
2    Africa                 Benin  1980            0  Production  Neuclear
3    Africa              Botswana  1980            0  Production  Neuclear
4    Africa          Burkina Faso  1980            0  Production  Neuclear

In [45]:
# print data type of all the columns
final_df.dtypes

Continent    object
Country      object
Year         object
Energy       object
Usage        object
Source       object
dtype: object

#### Finding time gaps

Year is datetime. I will convert it from object to datatime.
Energy is float number. I will convert it from object to numeric.

In [46]:
final_df['Year'] = final_df[['Year']].apply(pd.to_datetime)
#final_df['Year'] = pd.DatetimeIndex(final_df['Year']).year
#final_df['Year'] = pd.to_numeric(final_df['Year'], errors='coerce')  # 'coerce' will replace non-numeric values with NaN


In [47]:
final_df.dtypes

Continent            object
Country              object
Year         datetime64[ns]
Energy               object
Usage                object
Source               object
dtype: object

In [48]:
# Sorting dataframe by Year column
final_df = final_df.sort_values(by=['Year'])

In [51]:
# Creating a column with previous value of the year and subtracting one value ahead of it and storing result in Result column
#final_df['previous'] = (pd.DatetimeIndex(final_df['Year']).year).shift(1)
#final_df['Result'] = (pd.DatetimeIndex(final_df['Year']).year) - final_df['previous']
#print(final_df)

In [52]:
final_df[['Energy']] = final_df[['Energy']].apply(pd.to_numeric, errors='coerce')

In [53]:
# print data type of all the columns
final_df.dtypes

Continent            object
Country              object
Year         datetime64[ns]
Energy              float64
Usage                object
Source               object
dtype: object

In [54]:
# checking if a data can be selecteed using a continent 
final_df.loc[(final_df['Continent'] == 'Africa')]

Continent               Country       Year    Energy       Usage  \
0        Africa               Algeria 1980-01-01  0.002576  Production   
67666    Africa               Somalia 1980-01-01  0.000000  Production   
67667    Africa          South Africa 1980-01-01  0.000000  Production   
67668    Africa           South Sudan 1980-01-01       NaN  Production   
67669    Africa                 Sudan 1980-01-01  0.000000  Production   
...         ...                   ...        ...       ...         ...   
57757    Africa               Liberia 2021-01-01  0.000000  Production   
57758    Africa                 Libya 2021-01-01  2.632960  Production   
57759    Africa            Madagascar 2021-01-01  0.000000  Production   
57749    Africa              Ethiopia 2021-01-01  0.000000  Production   
57786    Africa              Zimbabwe 2021-01-01  0.000000  Production   

           Source  
0        Neuclear  
67666  NaturalGas  
67667  NaturalGas  
67668  NaturalGas  
67669  NaturalGas  
...           ...  
57757   Petrolium  
57758   Petrolium  
57759   Petrolium  
57749   Petrolium  
57786   Petrolium  

[19152 rows x 6 columns]

In [55]:
# checking if a data can be selecteed using a country 
final_df.loc[(final_df['Country'] == 'Algeria')]

Empty DataFrame
Columns: [Continent, Country, Year, Energy, Usage, Source]
Index: []

Selecting data using country didn't print anything, it means there could be white space or any other character around country value 

In [56]:
# print all the countries to check country column values
the_countries = set(final_df['Country'].to_list())
the_countries

{'        Afghanistan',
 '        Albania',
 '        Algeria',
 '        American Samoa',
 '        Angola',
 '        Antarctica',
 '        Antigua and Barbuda',
 '        Argentina',
 '        Armenia',
 '        Aruba',
 '        Australia',
 '        Austria',
 '        Azerbaijan',
 '        Bahrain',
 '        Bangladesh',
 '        Barbados',
 '        Belarus',
 '        Belgium',
 '        Belize',
 '        Benin',
 '        Bermuda',
 '        Bhutan',
 '        Bolivia',
 '        Bosnia and Herzegovina',
 '        Botswana',
 '        Brazil',
 '        British Virgin Islands',
 '        Brunei',
 '        Bulgaria',
 '        Burkina Faso',
 '        Burma',
 '        Burundi',
 '        Cabo Verde',
 '        Cambodia',
 '        Cameroon',
 '        Canada',
 '        Cayman Islands',
 '        Central African Republic',
 '        Chad',
 '        Chile',
 '        China',
 '        Colombia',
 '        Comoros',
 '        Congo-Brazzaville',
 '        Congo-Kinshasa'

There is white adding in front of country name. I will strip this white space for data cleaning purpose.

In [57]:
# strip white space of country column
final_df['Country'] = final_df['Country'].str.strip()

In [58]:
# check if data can be selecting using country name
final_df.loc[(final_df['Country'] == 'Algeria')]

Continent  Country       Year    Energy        Usage      Source
0        Africa  Algeria 1980-01-01  0.002576   Production    Neuclear
48300    Africa  Algeria 1980-01-01  2.315385   Production   Petrolium
67620    Africa  Algeria 1980-01-01  0.484980   Production  NaturalGas
9660     Africa  Algeria 1980-01-01  0.000076   Production        Coal
57960    Africa  Algeria 1980-01-01  0.542800  Consumption  NaturalGas
...         ...      ...        ...       ...          ...         ...
9430     Africa  Algeria 2021-01-01  0.007956   Production    Neuclear
19090    Africa  Algeria 2021-01-01  0.000000   Production        Coal
67390    Africa  Algeria 2021-01-01  1.767237  Consumption  NaturalGas
48070    Africa  Algeria 2021-01-01  0.000941  Consumption        Coal
57730    Africa  Algeria 2021-01-01  2.818340   Production   Petrolium

[336 rows x 6 columns]

#### Checking null values

In [59]:
final_df.isnull().sum()

Continent       0
Country         0
Year            0
Energy       5408
Usage           0
Source          0
dtype: int64

Energy column has 5408 missing values. 

#### Saving final dataframe to csv file

In [60]:
# saving final dataframe to csv file
final_df.to_csv('../Data_Files/Final_data.csv')

## 1.7 Summary<a id='1.7_Summary'></a>

To do energy production and consumpton forcasting, I needed data all the data in one dataframe with all the yearsin one column. To accomplish this goal, fist I check how data is in csv files, and then converting year rows to column using melt funtion and then explored data like column datatypes, selecting data based on a continent and cointry, checking if there are any missing values. Energy column has 5408 missing values. I will explore more about these missing values and how to deal with it in the next EDA notebook.